[BEAM-657] Support Read.Bounded primitive

amitsela opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

amitsela commented:
R: @dhalperi / @kennknowles.
Also @iemejia PTAL.

dhalperi commented:
R: @tgroh is a better standin for me in the short-term.

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+
+        @Override
+        public WindowedValue<T> next() {
+          return WindowedValue.of(reader.getCurrent(), reader.getCurrentTimestamp(),
```
`WindowedValue.timestampedValueInGlobalWindow`

amitsela replied:
Thanks.

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+
+        @Override
+        public void remove() {
+          // do nothing.
```
Is the runner expected to call this? If not, can probably throw an `UnsupportedOperationException` just to be more explicit about what the iterator supports, and/or to see when we're interacted with unexpectedly.

amitsela replied:
This will never be called, but indeed it's better to be explicit.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+      // when running local it's the total number of cores (local = 1, local[N] = N,
+      // local[*] = estimation of the machine's cores).
+      // ** the configuration "spark.default.parallelism" takes precedence over all of the above **
+      this.numPartitions = sc.defaultParallelism();
```
Funny enough, Spark won't assert `> 0`, so I will.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+      this.numPartitions = sc.defaultParallelism();
+    }
+
+    private static final long DEFAULT_BUNDLE_SIZE = 64 * 1024 * 1024;
```
HDFS default block size.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+  /**
+   * An input {@link Partition} wrapping the partitioned {@link Source}.
+   */
+  private static class SourcePartition<T> implements Partition {
```
I'm not too familiar with `AutoValue`, will take a look.

amitsela replied:
`AutoValue` seems like an overkill here, though I will change `hashCode()` a little bit to be more like Spark's other implementations like: https://github.com/apache/spark/blob/branch-1.6/core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala#L47

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:
```diff
@@ -561,10 +578,12 @@ public void evaluate(View.CreatePCollectionView<ReadT, WriteT> transform,
-    EVALUATORS.put(AvroIO.Read.Bound.class, readAvro());
-    EVALUATORS.put(AvroIO.Write.Bound.class, writeAvro());
+    //TODO: enable direct translations once BEAM-59 is resolved for hdfs (at least).
+//    EVALUATORS.put(TextIO.Read.Bound.class, readText());
```
Sounds fair.

amitsela replied:
To be handled by [BEAM-668](https://issues.apache.org/jira/browse/BEAM-668)

amitsela commented:
I'll take a look and have a second round of commits tomorrow. Thanks!

jbonofre commented:
R: @jbonofre

I noticed that source reader is not close (explicit call to `close()` method) if an exception occurs in the reader.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,199 @@
+              closeIfNotClosed();
+            }
+            return !finished;
+          } catch (IOException e) {
```
True.

amitsela commented:
@tgroh I addressed your comments, PTAL.
As for `RunnableOnService`, the SparkRunner was setup with it for a while and the only exclusion is `org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest`.
Somehow, it stopped running a while ago, more details here: [BEAM-259](https://issues.apache.org/jira/browse/BEAM-259)

amitsela commented:
Also, @jasonkuster@google.com (still not sure which annotation you respond to 😉 ) PTAL at the changes I made in `examples` - the SparkRunner should be treated as equal now 😄 .

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+import org.apache.spark.rdd.RDD;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Extra line

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+      // local[*] = estimation of the machine's cores).
+      // ** the configuration "spark.default.parallelism" takes precedence over all of the above **
+      this.numPartitions = sc.defaultParallelism();
+      checkArgument(this.numPartitions > 0, "Number of partitions must be greater than zero.");
```
Worth adding where the parallelism comes from (so configuration can be updated, if required)

amitsela replied:
Well, it comes from the `SparkContext`, how it determines it - that's Spark's business.
Spark requires set all kinds of resource-related settings, it's not reasonable to validate them all.
If you run local/YARN/Mesos then you're probably aware of that, and if you explicitly mentioned parallelism conf. you're aware as well.

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+        desiredSizeBytes = source.getEstimatedSizeBytes(
+            runtimeContext.getPipelineOptions()) / numPartitions;
+      } catch (Exception e) {
+        LOG.warn("Failed to get estimated size of bundle, default is {} bytes.",
```
"Failed to get estimated size of source {}, using default of {} bytes"?

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+        }
+        return partitions;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to create partitions.", e);
```
"...for Source %s" (maybe "for Source of type " + source.getClass().getSimpleName())

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+
+    @Override
+    public int hashCode() {
+      return 41 * (41 + rddId) + index;
```
Objects.hash()?

amitsela replied:
I'm really just trying to conform with Spark's different implementations of `hashCode()` in different implementations of `Partition` - and this is the way the went with implementing for "Hadoop" (which also includes s3, HBase, etc.). See: https://github.com/apache/spark/blob/branch-1.6/core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala#L52.

tgroh commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java:
```diff
@@ -0,0 +1,201 @@
+
+    @Override
+    public boolean equals(Object other) {
+      return super.equals(other);
```
According to your implementation of `hashCode` this should be
```
if (!(other instanceof SourcePartition)) {
  return false;
}
SourcePartition that = (SourcePartition) other;
return this.rddId == that.rddId && this.index == that.index
```

Which probably also means that you should include the `Source` in the equals and hash code methods

amitsela replied:
I'm actually removing `equals` as it is not required by `Partition`.

tgroh commented:
LGTM

@dhalperi @kennknowles @jbonofre for committers

kennknowles commented:
We've chatted, and my desire that "you should be able to run the local tests" is satisfied. We uncovered some issues, which is expected since the tests haven't been running. We can resolve them asynchronously, via other PRs and/or longer term tickets.

LGTM
